### PR TITLE
[clang] Remove an incorrect assertion in ConstantFoldAttrs

### LINKED ISSUE #119125: [clang] Assertion `Eval.Val.hasValue()' failed.
**This testcase is generated by a fuzzer.**

Compiler Explorer: [https://godbolt.org/z/zhK3Eddqs](https://godbolt.org/z/zhK3Eddqs)

This valid code will crash on clang assertion trunk:
```cpp
template<typename T>
struct B {
  [[clang::annotate("test", (9, (void)T{}))]] void t() {}
};
B<int> b;

```

This goes back to clang-12.

Stack dump:
```
clang++: /root/llvm-project/clang/lib/Sema/SemaAttr.cpp:540: bool clang::Sema::ConstantFoldAttrArgs(const clang::AttributeCommonInfo&, llvm::MutableArrayRef<clang::Expr*>): Assertion `Eval.Val.hasValue()' failed.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
0.	Program arguments: /opt/compiler-explorer/clang-assertions-trunk/bin/clang++ -gdwarf-4 -g -o /app/output.s -mllvm --x86-asm-syntax=intel -fno-verbose-asm -S --gcc-toolchain=/opt/compiler-explorer/gcc-snapshot -fcolor-diagnostics -fno-crash-diagnostics <source>
1.	<source>:5:9: current parser token ';'
2.	<source>:2:8: instantiating class definition 'B<int>'
 #0 0x0000000003c04df8 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+++0x3c04df8)
 #1 0x0000000003c02b04 llvm::sys::CleanupOnSignal(unsigned long) (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+++0x3c02b04)
 #2 0x0000000003b50068 CrashRecoverySignalHandler(int) CrashRecoveryContext.cpp:0:0
 #3 0x00007536fea42520 (/lib/x86_64-linux-gnu/libc.so.6+0x42520)
 #4 0x00007536fea969fc pthread_kill (/lib/x86_64-linux-gnu/libc.so.6+0x969fc)
 #5 0x00007536fea42476 gsignal (/lib/x86_64-linux-gnu/libc.so.6+0x42476)
 #6 0x00007536fea287f3 abort (/lib/x86_64-linux-gnu/libc.so.6+0x287f3)
 #7 0x00007536fea2871b (/lib/x86_64-linux-gnu/libc.so.6+0x2871b)
 #8 0x00007536fea39e96 (/lib/x86_64-linux-gnu/libc.so.6+0x39e96)
 #9 0x000000000673c949 clang::Sema::ConstantFoldAttrArgs(clang::AttributeCommonInfo const&, llvm::MutableArrayRef<clang::Expr*>) (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+++0x673c949)
#10 0x00000000066bfad7 clang::Sema::CreateAnnotationAttr(clang::AttributeCommonInfo const&, llvm::StringRef, llvm::MutableArrayRef<clang::Expr*>) (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+++0x66bfad7)
#11 0x00000000070db171 clang::Sema::InstantiateAttrs(clang::MultiLevelTemplateArgumentList const&, clang::Decl const*, clang::Decl*, llvm::SmallVector<clang::Sema::LateInstantiatedAttribute, 16u>*, clang::LocalInstantiationScope*) (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+++0x70db171)
#12 0x00000000070dc60c clang::TemplateDeclInstantiator::InitFunctionInstantiation(clang::FunctionDecl*, clang::FunctionDecl*) (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+++0x70dc60c)
#13 0x00000000070dcba8 clang::TemplateDeclInstantiator::InitMethodInstantiation(clang::CXXMethodDecl*, clang::CXXMethodDecl*) (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+++0x70dcba8)
#14 0x00000000070e2905 clang::TemplateDeclInstantiator::VisitCXXMethodDecl(clang::CXXMethodDecl*, clang::TemplateParameterList*, clang::TemplateDeclInstantiator::RewriteKind) (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+++0x70e2905)
#15 0x0000000007052e0b clang::Sema::InstantiateClass(clang::SourceLocation, clang::CXXRecordDecl*, clang::CXXRecordDecl*, clang::MultiLevelTemplateArgumentList const&, clang::TemplateSpecializationKind, bool) (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+++0x7052e0b)
#16 0x000000000707fd58 clang::Sema::InstantiateClassTemplateSpecialization(clang::SourceLocation, clang::ClassTemplateSpecializationDecl*, clang::TemplateSpecializationKind, bool) (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+++0x707fd58)
#17 0x00000000070f3f3f void llvm::function_ref<void ()>::callback_fn<clang::Sema::RequireCompleteTypeImpl(clang::SourceLocation, clang::QualType, clang::Sema::CompleteTypeKind, clang::Sema::TypeDiagnoser*)::'lambda'()>(long) SemaType.cpp:0:0
#18 0x0000000007c830f1 clang::StackExhaustionHandler::runWithSufficientStackSpace(clang::SourceLocation, llvm::function_ref<void ()>) (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+++0x7c830f1)
#19 0x00000000070fd367 clang::Sema::RequireCompleteTypeImpl(clang::SourceLocation, clang::QualType, clang::Sema::CompleteTypeKind, clang::Sema::TypeDiagnoser*) (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+++0x70fd367)
#20 0x00000000070fd6b5 clang::Sema::RequireCompleteType(clang::SourceLocation, clang::QualType, clang::Sema::CompleteTypeKind, clang::Sema::TypeDiagnoser&) (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+++0x70fd6b5)
#21 0x00000000070fd835 clang::Sema::RequireCompleteType(clang::SourceLocation, clang::QualType, clang::Sema::CompleteTypeKind, unsigned int) (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+++0x70fd835)
#22 0x00000000068d4a11 clang::Sema::ActOnUninitializedDecl(clang::Decl*) (.part.0) SemaDecl.cpp:0:0
#23 0x0000000006576928 clang::Parser::ParseDeclarationAfterDeclaratorAndAttributes(clang::Declarator&, clang::Parser::ParsedTemplateInfo const&, clang::Parser::ForRangeInit*) (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+++0x6576928)
#24 0x0000000006586409 clang::Parser::ParseDeclGroup(clang::ParsingDeclSpec&, clang::DeclaratorContext, clang::ParsedAttributes&, clang::Parser::ParsedTemplateInfo&, clang::SourceLocation*, clang::Parser::ForRangeInit*) (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+++0x6586409)
#25 0x00000000065463be clang::Parser::ParseDeclOrFunctionDefInternal(clang::ParsedAttributes&, clang::ParsedAttributes&, clang::ParsingDeclSpec&, clang::AccessSpecifier) (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+++0x65463be)
#26 0x0000000006546b79 clang::Parser::ParseDeclarationOrFunctionDefinition(clang::ParsedAttributes&, clang::ParsedAttributes&, clang::ParsingDeclSpec*, clang::AccessSpecifier) (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+++0x6546b79)
#27 0x000000000654e2f3 clang::Parser::ParseExternalDeclaration(clang::ParsedAttributes&, clang::ParsedAttributes&, clang::ParsingDeclSpec*) (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+++0x654e2f3)
#28 0x000000000654f1bd clang::Parser::ParseTopLevelDecl(clang::OpaquePtr<clang::DeclGroupRef>&, clang::Sema::ModuleImportState&) (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+++0x654f1bd)
#29 0x00000000065416fa clang::ParseAST(clang::Sema&, bool, bool) (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+++0x65416fa)
#30 0x0000000004587678 clang::CodeGenAction::ExecuteAction() (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+++0x4587678)
#31 0x0000000004842f09 clang::FrontendAction::Execute() (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+++0x4842f09)
#32 0x00000000047c2b9e clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+++0x47c2b9e)
#33 0x000000000492df6e clang::ExecuteCompilerInvocation(clang::CompilerInstance*) (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+++0x492df6e)
#34 0x0000000000cb235f cc1_main(llvm::ArrayRef<char const*>, char const*, void*) (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+++0xcb235f)
#35 0x0000000000ca9dfa ExecuteCC1Tool(llvm::SmallVectorImpl<char const*>&, llvm::ToolContext const&) driver.cpp:0:0
#36 0x00000000045cb0d9 void llvm::function_ref<void ()>::callback_fn<clang::driver::CC1Command::Execute(llvm::ArrayRef<std::optional<llvm::StringRef>>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>*, bool*) const::'lambda'()>(long) Job.cpp:0:0
#37 0x0000000003b50514 llvm::CrashRecoveryContext::RunSafely(llvm::function_ref<void ()>) (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+++0x3b50514)
#38 0x00000000045cb6cf clang::driver::CC1Command::Execute(llvm::ArrayRef<std::optional<llvm::StringRef>>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>*, bool*) const (.part.0) Job.cpp:0:0
#39 0x000000000459115d clang::driver::Compilation::ExecuteCommand(clang::driver::Command const&, clang::driver::Command const*&, bool) const (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+++0x459115d)
#40 0x000000000459224d clang::driver::Compilation::ExecuteJobs(clang::driver::JobList const&, llvm::SmallVectorImpl<std::pair<int, clang::driver::Command const*>>&, bool) const (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+++0x459224d)
#41 0x0000000004599605 clang::driver::Driver::ExecuteCompilation(clang::driver::Compilation&, llvm::SmallVectorImpl<std::pair<int, clang::driver::Command const*>>&) (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+++0x4599605)
#42 0x0000000000caf1a3 clang_main(int, char**, llvm::ToolContext const&) (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+++0xcaf1a3)
#43 0x0000000000b82fe4 main (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+++0xb82fe4)
#44 0x00007536fea29d90 (/lib/x86_64-linux-gnu/libc.so.6+0x29d90)
#45 0x00007536fea29e40 __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x29e40)
#46 0x0000000000ca98a5 _start (/opt/compiler-explorer/clang-assertions-trunk/bin/clang+++0xca98a5)
clang++: error: clang frontend command failed with exit code 134 (use -v to see invocation)
Compiler returned: 134
```

### PR DESCRIPTION
**Pull request by tbaederr**
Evaluating the attribute expression can be successful without resulting in a value. Namely, when the expression is of type void.

Fixes https://github.com/llvm/llvm-project/issues/119125

### CODE REVIEW

**Comment by tbaederr** [comment]
I guess it might make sense to create the `ConstantExpr` in any case, even for a `None` `APValue`? Dumping the `ConstantExpr` didn't print any `APValue` though, so I was assuming that it ignores `None` anyway.

**Comment by Sirraide** [comment]
> I guess it might make sense to create the `ConstantExpr` in any case, even for a `None` `APValue`?

The comment above `ConstantFoldAttrArgs` seems to suggest that that’s what’s supposed to happen if the expression isn’t dependent—I don’t know if there’s any code that depends on that, though.

**Comment by tbaederr** [comment]
> > I guess it might make sense to create the `ConstantExpr` in any case, even for a `None` `APValue`?
> 
> The comment above `ConstantFoldAttrArgs` seems to suggest that that’s what’s supposed to happen if the expression isn’t dependent—I don’t know if there’s any code that depends on that, though.

Hm, what comment are you referring to exactly?

We could of course also just reject void expressions, that's closer to the previous behavior I guess.

**Comment by Sirraide** [comment]
> Hm, what comment are you referring to exactly?

Ah, this one:
https://github.com/llvm/llvm-project/blob/d88876e74f7882643546becc544a771a5e5e9787/clang/include/clang/Sema/Sema.h#L1838-L1842

The main thing I’m (if only a bit) worried about is that some code somewhere might be expecting a `ConstantExpr` to always be there if the thing isn’t dependent, so an ‘empty’ `ConstantExpr` might be necessary if that’s the case, but then again, if everything works then it’s probably fine?

> We could of course also just reject void expressions

I mean, I think accepting them is fine (some attributes might have a use for them maybe?); just wondering how to best represent them in the AST.

**Comment by tbaederr** [comment]
CC @AaronBallman for an opinion about the AST represenation

**Comment by AaronBallman** [comment]
> CC @AaronBallman for an opinion about the AST represenation

This is a bit awkward in that it's only used for the `annotate` attributes' argument lists, which should pass all of their arguments down to LLVM IR. However, we don't typically generate LLVM IR for discarded value statements in other circumstances. So I'm not certain *what* we should pass down to LLVM IR in this case. The arguments could be used positionally, so in that case we'd want to pass *something* down so the arguments line up to the source. The arguments could be used semantically, so in that case, there's nothing to pass down because there are no semantics.

(CC @erichkeane for more opinions)

**Comment by erichkeane** [comment]
> > CC @AaronBallman for an opinion about the AST represenation
> 
> This is a bit awkward in that it's only used for the `annotate` attributes' argument lists, which should pass all of their arguments down to LLVM IR. However, we don't typically generate LLVM IR for discarded value statements in other circumstances. So I'm not certain _what_ we should pass down to LLVM IR in this case. The arguments could be used positionally, so in that case we'd want to pass _something_ down so the arguments line up to the source. The arguments could be used semantically, so in that case, there's nothing to pass down because there are no semantics.
> 
> (CC @erichkeane for more opinions)

My immediate read of this before seeing the discussion was that the positional arguments are important, so making one 'disappear' could be problematic.  I think a value of 'None', plus a reasonable IR for that would be our best way forward.

As far as what a 'reasonable IR' for a 'None' value would be, I don't know if we have any viable experts here we can discuss this with.  The nature of 'annotate' is that we HAVE no expert on the feature (since it is intended to be used by just about anyone in the backend).  I think we'd need to see a codegen test for 'none' however.

So in short:

1- Make it a None ConstantExpr

2- Give that a reasonable output for its attribute in IR

3- Write a codegen test that shows it off

 4- Add a release note.

**Comment by tbaederr** [comment]
For
```c++
void f() {
  [[clang::annotate("B", (void)1, (void)2, 7)]] int j = 0;
}
```

The IR is:
```llvm
@.str = private unnamed_addr constant [2 x i8] c"B\00", section "llvm.metadata"
@.str.1 = private unnamed_addr constant [10 x i8] c"array.cpp\00", section "llvm.metadata"
@.args = private unnamed_addr constant { i8, i8, i32 } { i8 undef, i8 undef, i32 7 }, section "llvm.metadata

; Function Attrs: mustprogress noinline nounwind optnone uwtable
define dso_local void @_Z1fv() #0 {
entry:
  %j = alloca i32, align 4
  call void @llvm.var.annotation.p0.p0(ptr %j, ptr @.str, ptr @.str.1, i32 187, ptr @.args)
  store i32 0, ptr %j, align 4
  ret void
}
```
So the void arguments get translated to `undef`. Is that acceptable?

**Comment by AaronBallman** [comment]
> For
> 
> ```c++
> void f() {
>   [[clang::annotate("B", (void)1, (void)2, 7)]] int j = 0;
> }
> ```
> 
> The IR is:
> 
> ```llvm
> @.str = private unnamed_addr constant [2 x i8] c"B\00", section "llvm.metadata"
> @.str.1 = private unnamed_addr constant [10 x i8] c"array.cpp\00", section "llvm.metadata"
> @.args = private unnamed_addr constant { i8, i8, i32 } { i8 undef, i8 undef, i32 7 }, section "llvm.metadata
> 
> ; Function Attrs: mustprogress noinline nounwind optnone uwtable
> define dso_local void @_Z1fv() #0 {
> entry:
>   %j = alloca i32, align 4
>   call void @llvm.var.annotation.p0.p0(ptr %j, ptr @.str, ptr @.str.1, i32 187, ptr @.args)
>   store i32 0, ptr %j, align 4
>   ret void
> }
> ```
> 
> So the void arguments get translated to `undef`. Is that acceptable?

That's an interesting situation! I'm not sure what I'd expect the LLVM IR to be there. For Clang AST, we want to keep around the expression AST nodes. But for LLVM IR, perhaps it makes sense to drop anything without a value rather than mark it as undef?

**Comment by nikic** [comment]
> That's an interesting situation! I'm not sure what I'd expect the LLVM IR to be there. For Clang AST, we want to keep around the expression AST nodes. But for LLVM IR, perhaps it makes sense to drop anything without a value rather than mark it as undef?

No strong opinion, I think either way would be okay. Using undef has the very slight advantage that there's still an indication that there was an argument there, even if it doesn't have a value, so `[[clang::annotation("a", (void)1, 2)]]` and `[[clang::annotation("a", 2)]]` can be distinguished.

**Comment by tbaederr** [comment]
Sounds like an LGTM from Nikita, does anyone else have other reservations? (Other than the missing release note)

**Review by AaronBallman — APPROVED** [review]
LGTM aside from the missing release note.
